clarity-ext-scripts/clarity_ext_scripts/covid/fetch_biobank_barcodes.py

```diff
@@ -0,0 +1,137 @@
+
+        # Validate that all wells in biobank file are matched in sample list file
+        file_stream2 = self.context.local_shared_file('Raw sample list')
+        sample_matrix = self._build_sample_matrix(file_stream2, plate_barcode)
+        for key in biobank_matrix:
+            try:
+                if biobank_matrix[key]['biobank_barcode'] != 'NO TUBE':
+                    _ = sample_matrix[key]['Sample Id']
+            except KeyError:
+                raise UsageError(
+                    "The well references are not matching between the "
+                    "files in 'Raw biobank list' and 'Raw sample list, "
+                    "well '{}' has no match in the 'Raw sample list'"
+                    .format(biobank_matrix[key]['well'])
+                )
+
+        # Validate that 'NO TUBE' entries in biobank file is empty in sample list
+        # Perhaps this is a little too aggressive validation?
+        sample_matrix_keys = [k for k in sample_matrix]
+        for key in biobank_matrix:
+            if biobank_matrix[key]['biobank_barcode'] == 'NO TUBE' \
+                    and key in sample_matrix_keys \
+                    and sample_matrix[key]['Sample Id']:
+                biobank_well = biobank_matrix[key]['well']
+                sample_list_well = sample_matrix[key]['Position']
+                raise UsageError(
+                    "There is an empty entry in the biobank barcode file "
+                    "that is not empty in the sample list file, "
+                    "biobank well: {}, sample list well: {}"
+                    .format(biobank_well, sample_list_well))
+
+    def biobank_barcode_by_sample_referal_code(self):
+        file_stream = self.context.local_shared_file('Raw biobank list')
+        biobank_matrix = self._build_biobank_matrix(file_stream)
+        plate_barcode = self._plate_barcode_from(biobank_matrix)
+        file_stream2 = self.context.local_shared_file('Raw sample list')
+        sample_matrix = self._build_sample_matrix(file_stream2, plate_barcode)
+        barcode_map = dict()
+        for key in biobank_matrix:
+            if biobank_matrix[key]['biobank_barcode'] == 'NO TUBE':
+                continue
+            biobank_barcode = biobank_matrix[key]['biobank_barcode']
+            sample_referal_code = sample_matrix[key]['Sample Id']
+            barcode_map[sample_referal_code] = biobank_barcode
+
+        return barcode_map
+
+    def _build_sample_matrix(self, file_stream, plate_barcode):
+        csv = Csv(file_stream)
+        matrix = dict()
+        pattern = re.compile(r"(?P<row>[A-Z])(?P<col>[0-9]+)")
+        for line in csv:
+            trimmed_row = map(str.strip, line.values)
+            row_as_dict = dict(zip(csv.header, trimmed_row))
+            well_robot_format = line['Position']
+            m = pattern.match(well_robot_format)
+            if m is None:
+                continue
```

Comment:
Is it OK to silently do nothing if the position is not matched? If it's not expected to happen, I would prefer an exception to be thrown.

Reply:
It does happen because the end of the file is different. So this was a quick and dirty way to fix the different format at end. Should I perhaps do a little bit more extensive check on the file format?

Reply:
Ahh I see! I don't know, I think just commenting on that would be good too. It might also be clearer to just break entirely out of it in that case, since I guess you don't expect it to start containing valid positions after it stops returning them.